pkg/antctl/raw/migrate/convertnetworkpolicy/command.go

```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		klog.Errorf("Failed to list Calico NetworkSets: %v", err)
+		return err
+	}
+	for _, ns := range nss.Items {
+		klog.Infof("Converting NetworkSet %s", ns.Name)
+		group := v1beta1.Group{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ns.Name,
+				Namespace: ns.Namespace,
+			},
+			Spec: v1beta1.GroupSpec{
+				IPBlocks: []v1beta1.IPBlock{},
+			},
+		}
+
+		for _, cidr := range ns.Spec.Nets {
+			group.Spec.IPBlocks = append(group.Spec.IPBlocks, v1beta1.IPBlock{CIDR: cidr})
+		}
+		if _, err := c.antreaClient.CrdV1beta1().Groups(ns.Namespace).Create(context.Background(), &group, metav1.CreateOptions{}); err != nil {
+			klog.Errorf("Failed to create Antrea Group %s: %v", ns.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *converter) convertGlobalNetworkSet() error {
+	gnss, err := c.calicoClient.ProjectcalicoV3().GlobalNetworkSets().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Failed to list Calico GlobalNetworkSets: %v", err)
+		return err
+	}
+	for _, gns := range gnss.Items {
+		klog.Infof("Converting GlobalNetworkSet %s", gns.Name)
+		clusterGroup := v1beta1.ClusterGroup{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: gns.Name,
+			},
+			Spec: v1beta1.GroupSpec{
+				IPBlocks: []v1beta1.IPBlock{},
+			},
+		}
+
+		for _, cidr := range gns.Spec.Nets {
+			clusterGroup.Spec.IPBlocks = append(clusterGroup.Spec.IPBlocks, v1beta1.IPBlock{CIDR: cidr})
+		}
+		if _, err := c.antreaClient.CrdV1beta1().ClusterGroups().Create(context.Background(), &clusterGroup, metav1.CreateOptions{}); err != nil {
+			klog.Errorf("Failed to create Antrea Group %s: %v", gns.Name, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *converter) convertCalicoNamespacedNetworkPolicy() error {
+	nps, err := c.calicoClient.ProjectcalicoV3().NetworkPolicies("").List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("Failed to list Calico NetworkPolicies: %v", err)
+		return err
+	}
+	for _, np := range nps.Items {
+		klog.Infof("Converting NetworkPolicy %s", np.Name)
+		anp := v1beta1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      np.Name,
+				Namespace: np.Namespace,
+				Annotations: map[string]string{
+					convertFromCalicoAnnotation: "true",
+				},
+			},
+		}
+
+		if anp.Spec.AppliedTo, err = convertSelectorToAntreaAppliedTo(np.Spec.Selector, ""); err != nil {
+			return err
+		}
+		if anp.Spec.Egress, err = convertCalicoRuleToAntreaRule(np.Spec.Egress); err != nil {
+			return err
+		}
+		if anp.Spec.Ingress, err = convertCalicoRuleToAntreaRule(np.Spec.Ingress); err != nil {
+			return err
+		}
+		if np.Spec.Order != nil {
+			anp.Spec.Priority = *np.Spec.Order
+		} else {
+			anp.Spec.Priority = defaultPriority
+		}
+		if _, err := c.antreaClient.CrdV1beta1().NetworkPolicies(np.Namespace).Create(context.Background(), &anp, metav1.CreateOptions{}); err != nil {
```

Comment:
Is this converter idempotent? should we retry here if it's a retryable error? or let user to rerun the command?

Reply:
Actually, if the migration process fails, all Antrea CRDs will be deleted. So every time we run this conversion, it will be idempotent, but in case of some stale resources, I added a check for the `IsAlreadyExist` error.